### PR TITLE
[Feature] Unordered parallel graph walk V2

### PR DESCRIPTION
**Pull request by ruseinov**
## Summary of changes

Changes introduced in this pull request:

- Introduced unordered parallel graph traversal to speed up operations that don't care about the block order.

## Reference issue to close (if applicable)

Work on #3314 

## Change checklist

- [ ] I have performed a self-review of my own code,
- [ ] I have made corresponding changes to the documentation,
- [ ] I have added tests that prove my fix is effective or that my feature works (if possible),
- [ ] I have made sure the [CHANGELOG][1] is up-to-date. All user-facing changes should be reflected in this document.

[1]: https://github.com/ChainSafe/forest/blob/main/CHANGELOG.md

### CODE REVIEW

File: src/ipld/util.rs [hunk 1]
```diff
@@ -438,3 +447,233 @@ impl<DB: Blockstore, T: Iterator<Item = Tipset> + Unpin> Stream for ChainStream<
...
+                let db = db.clone();
+                let block_sender = block_sender.clone();
+                handles.spawn(async move {
+                    while let Ok(mut cid_vec) = extract_receiver.recv_async().await {
+                        while let Some(cid) = cid_vec.pop() {
+                            if should_save_block_to_snapshot(cid) && seen.lock().insert(cid) {
+                                if let Some(data) = db.get(&cid)? {
+                                    if cid.codec() == fvm_ipld_encoding::DAG_CBOR {
+                                        let mut new_values = extract_cids(&data)?;
+                                        cid_vec.reserve(new_values.len());
```
Comment (lemmih) [inline src/ipld/util.rs:548]: `Vec::append` already calls reserve.

Comment (ruseinov) [reply src/ipld/util.rs:548]: Yep, this use to push in a loop, needs to be removed. Thanks!

**Comment by ruseinov** [comment]
This looks good. It's still quite slow at one point.  Checking `unordered without has` now. 
```
unordered with has                     traversed 87.43 GiB at 387.18 MiB/s in 00:03:51
unordered without parallel part traversed 13.70 GiB at 252.89 MiB/s in 00:00:55
ordered                                        traversed 87.43 GiB at 214.33 MiB/s in 00:06:57
ordered only sequential part     traversed 13.70 GiB at 257.50 MiB/s in 00:00:54
```

**Comment by ruseinov** [comment]
Closed in favour of #3474

**Pull request closed without merging**